cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking.qll

```diff
@@ -157,6 +157,15 @@ module TaintTracking {
     }
   }
 
+  /**
+   * Holds if there is local taint flow from parameter `src` to parameter `dest`
+   * of function `f`.
+   */
+  private predicate parameterTaintFlow(Function f, int src, int dest) {
+    TaintTracking::localTaint(DataFlow::parameterNode(f.getParameter(src)),
+      DataFlow::exprNode(f.getParameter(dest).getAnAssignedValue()))
+  }
```

Comment:
1. Why is flow between arguments classified as taint and not data flow? If it preserves the value perfectly and doesn't conflate objects with pointers, I'd say it's normal data flow.
2. Shouldn't `parameterTaintFlow` only hold if the destination parameter is of reference type? Otherwise I'd expect false flow in `void myNotAssign2(int *a, int *b) { a = b; }`.
3. This implementation works by recursion through a transitive closure (`localTaint`). That's not how the rest of the library works, and I'm concerned about performance.
4. You've added this interprocedural step in `localTaintStep`, but "local" here means that it's not supposed to be interprocedural. This is related to 3. I think the right place for this code is in the shared interprocedural data flow library that's also used by C# and Java. There's now a "return kind" concept that lets a function return multiple values, and it's used by C# for `out` parameters. This is related to our current efforts to get flow through fields.

Reply:
> shared interprocedural data flow library

Where should I look for this?

There's a sense in which this isn't interprocedural, because both the source and destination are at the call site, and these patterns sometimes mirror primitive operations that we do support with local data flow.  For example:
```
a = b; // data flows from b to a
memcpy(&a, &b, sizeof(a)); // data flows from b to a
```
However I don't think I have particularly strong feelings on this.

> I'm concerned about performance.

Yes ... I'll look into this.

Reply:
I took this discussion with the other data flow maintainers in Copenhagen, and the consensus was that adding local flow (or taint) through callee bodies would go against the main application of local flow: finding conservative flow that isn't likely to have false positives caused by interprocedural confusion (including virtual dispatch in the case of Java and C#). When we model a library function like `memcpy`, we consider whether the flow we're adding is likely to correspond with real-world flow; auto-detection of flow cannot do that unless we add an "underapproximation mode" to it.

I think the right way to make this change is to support "out parameters" like in C#. Compare https://github.com/Semmle/ql/blob/d0a761477b069bf0f255e37990a7f1f4b3e19684/cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowPrivate.qll#L53 to https://github.com/Semmle/ql/blob/1c25e17812a78306eb57235e07ec4ba4320ef2e1/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowDispatch.qll#L74-L87.

That's a pretty big change, and it could conflict and overlap with the steps we're taking to add flow through fields. I suggest that we postpone this PR until the design for flow through fields has settled. In the mean time, the potential customer who asked for this could get a separate "local flow with flow through params" predicate, either in the standard library or in their own QL repo. That predicate could be deprecated after flow through fields is released. If you're adding flow from in-parameters to out-parameters, I think it would be consistent and useful to also add flow from in-parameters to return values.